proof/refine/AARCH64/ArchInvsLemmas_H.thy

```diff
@@ -3449,99 +327,171 @@ lemma range_cover_canonical_address:
+                   typ_at_lift_ntfn' typ_at_lift_cte_at'
+                   hoare_vcg_conj_lift [OF typ_at_lift_cte_at'])
+     apply (rename_tac zombie_type nat)
+     apply (case_tac zombie_type; simp)
+      apply (wp typ_at_lift_tcb' P hoare_vcg_all_lift typ_at_lift_cte')+
+    apply (rename_tac arch_capability)
+    apply (case_tac arch_capability,
+           simp_all add: P[of id, simplified] vspace_table_at'_defs
+                         hoare_vcg_prop All_less_Ball
+                    split del: if_split)
+       apply (wp hoare_vcg_const_Ball_lift P typ_at_lift_valid_untyped'
+                 hoare_vcg_all_lift typ_at_lift_cte' typ_at_lift_frame_at')+
+  done
 
-add_upd_simps "invs' (gsUntypedZeroRanges_update f s)"
-  (obj_at'_real_def)
-declare upd_simps[simp]
-
-lemma neq_out_intv:
-  "\<lbrakk> a \<noteq> b; b \<notin> {a..a + c - 1} - {a} \<rbrakk> \<Longrightarrow> b \<notin> {a..a + c - 1}"
-  by simp
-
-lemma rule_out_intv:
-  "\<lbrakk> ksPSpace s a = Some obj; ksPSpace s b = Some obj'; pspace_distinct' s; a \<noteq> b \<rbrakk>
-   \<Longrightarrow> b \<notin> mask_range a (objBitsKO obj)"
-  apply (drule(1) pspace_distinctD')
-  apply (subst (asm) ps_clear_def)
-  apply (drule_tac x = b in orthD2)
-   apply fastforce
-  apply (drule neq_out_intv)
-   apply (simp add: mask_def add_diff_eq)
-  apply (simp add: mask_def add_diff_eq)
+lemma valid_arch_tcb_lift':
+  assumes x: "\<And>T p. \<lbrace>typ_at' T p\<rbrace> f \<lbrace>\<lambda>rv. typ_at' T p\<rbrace>"
+  shows "\<lbrace>\<lambda>s. valid_arch_tcb' tcb s\<rbrace> f \<lbrace>\<lambda>rv s. valid_arch_tcb' tcb s\<rbrace>"
+  apply (clarsimp simp add: valid_arch_tcb'_def)
+  apply (cases "atcbVCPUPtr tcb"; simp)
+   apply (wp x)+
   done
 
-lemma ptr_range_mask_range:
-  "{ptr..ptr + 2 ^ bits - 1} = mask_range ptr bits"
-  unfolding mask_def
-  by simp
-
-lemma distinct_obj_range'_not_subset:
-  "\<lbrakk> ksPSpace s a = Some obj; ksPSpace s b = Some obj'; pspace_distinct' s;
-     pspace_aligned' s; a \<noteq> b \<rbrakk>
-   \<Longrightarrow> \<not> obj_range' b obj' \<subseteq> obj_range' a obj"
-  unfolding obj_range'_def
-  apply (frule_tac x=a in pspace_alignedD')
-   apply assumption
-  apply (frule_tac x=b in pspace_alignedD')
-   apply assumption
-  apply (frule (3) rule_out_intv)
-  using is_aligned_no_overflow_mask
-  by fastforce
-
-lemma obj_range'_disjoint:
-  "\<lbrakk> ksPSpace s a = Some obj; ksPSpace s b = Some obj'; pspace_distinct' s;
-     pspace_aligned' s; a \<noteq> b \<rbrakk>
-   \<Longrightarrow> obj_range' a obj \<inter> obj_range' b obj' = {}"
-  apply (frule_tac x=a in pspace_alignedD')
-   apply assumption
-  apply (frule_tac x=b in pspace_alignedD')
-   apply assumption
-  apply (frule_tac p=a and p'=b in aligned_mask_range_cases)
-   apply assumption
-  apply (fastforce dest: distinct_obj_range'_not_subset
-                   simp: obj_range'_def)
+(* FIXME arch-split: do we want a gen_typ_at_lifts version? *)
```

Comment:
Would not be useful in the proofs, but might be useful in terms of making sure that each arch has all generic objects (in case we later ever add a generic object). So in terms of organising things, probably yes, but can be a FIXME for now.

Reply:
I managed to create one and put it in InvariantsUpdates_H. next round